File: lab4-mvc-engineer/src/main/scala/u04/mvc/engineer/SwingFunctionalFacade.java
package scala.u04.mvc.engineer;

import javax.swing.*;
import java.awt.*;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.function.Supplier;

class SwingFunctionalFacade {

    public interface Frame {
        Frame setSize(int width, int height);
        Frame addPanel(JPanel panel, String name);
        Frame drawEllipse(String panelName, int x, int y, int radius);
        Frame schedule(int millis, String eventName);
        Frame show();
        Supplier<String> events();        
    }

    public static Frame createFrame(){
        return new FrameImpl();
    }

    private static class FrameImpl implements Frame {
        private final JFrame jframe = new JFrame();
        private final Map<String, JPanel> jPanels = new HashMap<>();
        private final LinkedBlockingQueue<String> eventQueue = new LinkedBlockingQueue<>();
        private final Supplier<String> events = () -> {
            try{
                return eventQueue.take();
            } catch (InterruptedException e){
                return "";
            }
        };
        public FrameImpl() {
            this.jframe.setLayout(new GridLayout(1, 1));
            this.jframe.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        }

        @Override
        public Frame setSize(int width, int height) {
            this.jframe.setSize(width, height);
            return this;
        }

        @Override
        public Frame addPanel(JPanel panel, String name) {
            panel.setPreferredSize(this.jframe.getSize());
            panel.setBackground(Color.BLACK);
            this.jframe.add(panel);
            this.jPanels.put(name, panel);
            return this;
        }

        @Override
        public Frame drawEllipse(String panelName, int x, int y, int radius) {
            if (this.jPanels.containsKey(panelName)) {
                var panel = this.jPanels.get(panelName);
                if (panel instanceof SwingCustomComponents.BallPanel drawablePanel) {
                    drawablePanel.drawBall(x, y, radius);
                    drawablePanel.repaint();
                }
            }
            return this;
        }

        @Override
        public Frame schedule(int millis, String eventName) {
            Timer timer = new Timer(millis, (ev) -> {
                try {
                    eventQueue.put(eventName);
                } catch (InterruptedException ex){}
            });
            timer.setRepeats(false); 
            timer.start();
            return this;
        }


        @Override
        public Supplier<String> events() {
            return events;
        }

        @Override
        public Frame show() {
            this.jframe.setVisible(true);
            return this;
        }

    }
}

File: chemistLab7/src/main/scala/u07/simulation/SwingFunctionalFacade.scala
package u07.simulation;

import u04.mvc.engineer.SwingCustomComponents

import javax.swing.*
import javax.swing.WindowConstants.EXIT_ON_CLOSE
import java.awt.*
import java.util.concurrent.LinkedBlockingQueue
import java.util.function.Supplier
import scala.collection.mutable
import scala.util.Try

object SwingFunctionalFacade:

  def createFrame(): Frame = new FrameImpl();

  trait Frame:
    def setSize(width: Int, height: Int): Frame
    def addPanel(panel: JPanel, name: String): Frame
    def schedule(millis: Int, eventName: String): Frame
    def show(): Frame
    def events(): () => String


  class FrameImpl extends Frame:
    private val jFrame: JFrame = new JFrame()
    private val jPanels: mutable.Map[String, JPanel] = mutable.Map()
    private val eventQueue: LinkedBlockingQueue[String] = new LinkedBlockingQueue()
    override def events(): () => String = () =>
      Try:
        eventQueue.take()
      .getOrElse("")

    this.jFrame.setLayout(new GridLayout(1, 1))
    this.jFrame.setDefaultCloseOperation(EXIT_ON_CLOSE)

    override def setSize(width: Int, height: Int): Frame =
      this.jFrame.setSize(width, height)
      this

    override def addPanel(panel: JPanel, name: String): Frame =
      panel.setPreferredSize(this.jFrame.getSize());
      panel.setBackground(Color.BLACK);
      this.jFrame.add(panel);
      this.jPanels += (name -> panel);
      this

    def schedule(millis: Int, eventName: String): Frame = 
      val timer = new Timer(millis, _ => Try(eventQueue.put(eventName)))
      timer.setRepeats(false)
      timer.start()
      this
      
    def show(): Frame = 
      jFrame.setVisible(true)
      this



